sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py

```diff
@@ -124,26 +124,26 @@ def send(self, request: PipelineRequest[HTTPRequestType]) -> PipelineResponse[HT
         self.on_request(request)
         try:
             response = self.next.send(request)
-            self.on_response(request, response)
         except Exception:  # pylint:disable=broad-except
             self.on_exception(request)
             raise
-        else:
-            if response.http_response.status_code == 401:
-                self._token = None  # any cached token is invalid
-                if "WWW-Authenticate" in response.http_response.headers:
-                    request_authorized = self.on_challenge(request, response)
-                    if request_authorized:
-                        # if we receive a challenge response, we retrieve a new token
-                        # which matches the new target. In this case, we don't want to remove
-                        # token from the request so clear the 'insecure_domain_change' tag
-                        request.context.options.pop("insecure_domain_change", False)
-                        try:
-                            response = self.next.send(request)
-                            self.on_response(request, response)
-                        except Exception:  # pylint:disable=broad-except
-                            self.on_exception(request)
-                            raise
+
+        self.on_response(request, response)
```

Comment:
Can `self.on_response` raise an exception? If so `self.on_exception` is no longer called here. Is that intentional?

Reply:
It's a no-op for the default policy, but I guess it could if someone subclasses it to create a custom policy. We have a mismatch of where this is called in sync vs async - sync had it in the try/except and async had it outside.

Reply:
There was some discussion [here](https://github.com/Azure/azure-sdk-for-python/pull/31766/files#r1307987218) regarding the topic of `on_response` and `on_exception` being mutually exclusive when it was moved in the async policy. I think we should follow suit.